staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go

```diff
@@ -213,6 +241,20 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 		return nil, actual, err
 	}
 
+	// Handle strict mode.
+	if s.strict {
+		// Delayed failure for the strict YAML to JSON conversation,
+		// so that the error output also includes GVK.
+		if errorYAMLToJSON != nil {
+			return nil, actual, fmt.Errorf("%#v: %v", actual, errorYAMLToJSON)
```

Comment:
I expected a structured error we could pull the list of bad fields out of

Reply:
i think we need to define two error types
- one for duplicate fields
it uses yaml.2 as a backend and just has a slice of regular errors, without fields:
https://github.com/go-yaml/yaml/blob/v2.2.2/yaml.go#L271
https://github.com/go-yaml/yaml/blob/v2/decode.go#L645

- one for unknown fields:
this just uses encoding/json from stdlib, where sadly the error was added as untyped too and only provides text info for a the first failing field it encounters:
https://github.com/golang/go/blob/master/src/encoding/json/decode.go#L736

so the best we can do for now is define our types locally and possibly have them return a list of []errors. but we cannot be extra information in the errors (unless we want to parse Error() string output).

Reply:
most importantly, IMO, is to be able to know what problem there was on caller side, so declaring a new error type (e.g. `StrictUnmarshalError`) that then support some `.Cause()` (either "Unknown Fields" or "Duplicate Fields", as a specific string constant type) method would be a good start. If we then additionally can get the fields that's and extra plus but without some fancy string parsing it doesn't seem very easily doable

Reply:
...or maybe using two dedicated error types is better